## Decision Tree

### Credit Dataset
The goal of the project is to identify Whether it is a default or not with given features like checking_balance, months_loan_duration, credit_history etc using decision tree algorithm.
improving the performence using Boosting and penalizing. 
checking whether those methods really improving accuracy or not.

### Mashroom datset
The goal of the project is to identify whether the type of mashroom is edible or poisonoue with given feature.
Improving the performace with decision rule making.
comparing the models.

In [4]:
# imported required packages
install.packages("psych")
library(psych)
install.packages("tidyverse")
library(tidyverse)
install.packages("stringr")
library(stringr)
install.packages("DAAG")                
library(DAAG)
install.packages("ggplot2")                 
library(ggplot2)
install.packages("stats")                 
library(stats)
install.packages("gridExtra")                 
library(gridExtra)
install.packages("glmnet")                
library(glmnet)
install.packages("devtools")                
library(devtools)
install.packages("MASS")                 
library(MASS)
install.packages("data.table")
library(data.table)
install.packages("ggrepel")                 
library(ggrepel)
install.packages("lubridate")
library(lubridate)
install.packages("rgdal")                 
library(rgdal)                 
install.packages("ipred")
install.packages("caret")
library(caret)

also installing the dependency 'mnormt'



package 'mnormt' successfully unpacked and MD5 sums checked
package 'psych' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\kgajj\AppData\Local\Temp\RtmpGyFmrC\downloaded_packages
package 'tidyverse' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\kgajj\AppData\Local\Temp\RtmpGyFmrC\downloaded_packages


Registered S3 methods overwritten by 'ggplot2':
  method         from 
  [.quosures     rlang
  c.quosures     rlang
  print.quosures rlang
Registered S3 method overwritten by 'rvest':
  method            from
  read_xml.response xml2
-- Attaching packages --------------------------------------- tidyverse 1.2.1 --
v ggplot2 3.1.1     v readr   1.3.1
v tibble  2.1.1     v purrr   0.3.2
v tidyr   0.8.3     v stringr 1.4.0
v ggplot2 3.1.1     v forcats 0.4.0
-- Conflicts ------------------------------------------ tidyverse_conflicts() --
x ggplot2::%+%()   masks psych::%+%()
x ggplot2::alpha() masks psych::alpha()
x dplyr::filter()  masks stats::filter()
x dplyr::lag()     masks stats::lag()
Warning message:
"package 'stringr' is in use and will not be installed"also installing the dependency 'latticeExtra'



package 'latticeExtra' successfully unpacked and MD5 sums checked
package 'DAAG' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\kgajj\AppData\Local\Temp\RtmpGyFmrC\downloaded_packages


Loading required package: lattice

Attaching package: 'DAAG'

The following object is masked from 'package:psych':

    cities

Warning message:
"package 'ggplot2' is in use and will not be installed"Warning message:
"package 'stats' is not available (for R version 3.6.0)"Warning message:
"package 'stats' is a base package, and should not be updated"

package 'gridExtra' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\kgajj\AppData\Local\Temp\RtmpGyFmrC\downloaded_packages



Attaching package: 'gridExtra'

The following object is masked from 'package:dplyr':

    combine



package 'glmnet' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\kgajj\AppData\Local\Temp\RtmpGyFmrC\downloaded_packages


Loading required package: Matrix

Attaching package: 'Matrix'

The following object is masked from 'package:tidyr':

    expand

Loading required package: foreach

Attaching package: 'foreach'

The following objects are masked from 'package:purrr':

    accumulate, when

Loaded glmnet 2.0-18

also installing the dependencies 'ini', 'desc', 'rprojroot', 'xopen', 'clisymbols', 'gh', 'git2r', 'memoise', 'pkgbuild', 'pkgload', 'rcmdcheck', 'remotes', 'sessioninfo', 'usethis'



package 'ini' successfully unpacked and MD5 sums checked
package 'desc' successfully unpacked and MD5 sums checked
package 'rprojroot' successfully unpacked and MD5 sums checked
package 'xopen' successfully unpacked and MD5 sums checked
package 'clisymbols' successfully unpacked and MD5 sums checked
package 'gh' successfully unpacked and MD5 sums checked
package 'git2r' successfully unpacked and MD5 sums checked
package 'memoise' successfully unpacked and MD5 sums checked
package 'pkgbuild' successfully unpacked and MD5 sums checked
package 'pkgload' successfully unpacked and MD5 sums checked
package 'rcmdcheck' successfully unpacked and MD5 sums checked
package 'remotes' successfully unpacked and MD5 sums checked
package 'sessioninfo' successfully unpacked and MD5 sums checked
package 'usethis' successfully unpacked and MD5 sums checked
package 'devtools' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\kgajj\AppData\Local\Temp\RtmpGyFmrC\dow


Attaching package: 'MASS'

The following object is masked from 'package:DAAG':

    hills

The following object is masked from 'package:dplyr':

    select



package 'data.table' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\kgajj\AppData\Local\Temp\RtmpGyFmrC\downloaded_packages



Attaching package: 'data.table'

The following object is masked from 'package:purrr':

    transpose

The following objects are masked from 'package:dplyr':

    between, first, last



package 'ggrepel' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\kgajj\AppData\Local\Temp\RtmpGyFmrC\downloaded_packages
package 'lubridate' successfully unpacked and MD5 sums checked


Warning message:
"cannot remove prior installation of package 'lubridate'"


The downloaded binary packages are in
	C:\Users\kgajj\AppData\Local\Temp\RtmpGyFmrC\downloaded_packages


ERROR: Error in library(lubridate): there is no package called 'lubridate'


## Data Extraction

The dataset is extracted from Machine learning with R Git hub.Dataset contains 1000 record with 21 featues. In this default is the target variable. The default vector indicates whether the loan applicant was unable to meet the agreed payment terms and went into default which is highly undesirable.

In [12]:
# extracting Data
credit  <- read.csv("C:\\Users\\kgajj\\AppData\\Local\\Temp\\Temp1_Machine-Learning-with-R-datasets-master.zip\\Machine-Learning-with-R-datasets-master\\credit.csv")

In [8]:
#head of the dataset
head(credit)

checking_balance months_loan_duration credit_history purpose   amount
1 < 0 DM            6                   critical       radio/tv  1169  
2 1 - 200 DM       48                   repaid         radio/tv  5951  
3 unknown          12                   critical       education 2096  
4 < 0 DM           42                   repaid         furniture 7882  
5 < 0 DM           24                   delayed        car (new) 4870  
6 unknown          36                   repaid         education 9055  
  savings_balance employment_length installment_rate personal_status
1 unknown         > 7 yrs           4                single male    
2 < 100 DM        1 - 4 yrs         2                female         
3 < 100 DM        4 - 7 yrs         2                single male    
4 < 100 DM        4 - 7 yrs         2                single male    
5 < 100 DM        1 - 4 yrs         3                single male    
6 unknown         1 - 4 yrs         2                single male    
  other_debtors ... property                 age installment_plan housing 
1 none          ... real estate              67  none             own     
2 none          ... real estate              22  none             own     
3 none          ... real estate              49  none             own     
4 guarantor     ... building society savings 45  none             for free
5 none          ... unknown/none             53  none             for free
6 none          ... unknown/none             35  none             for free
  existing_credits default dependents telephone foreign_worker
1 2                1       1          yes       yes           
2 1                2       1          none      yes           
3 1                1       2          none      yes           
4 1                1       2          none      yes           
5 2                2       2          none      yes           
6 1                1       2          yes       yes           
  job               
1 skilled employee  
2 skilled employee  
3 unskilled resident
4 skilled employee  
5 skilled employee  
6 unskilled resident

## Descriptive statistics

In credit dataset,there are no missing values in the datset. The dataset looks clean.

In [9]:
#dimensions of the datset
dim(credit)

[1] 1000   21

In [10]:
# structure of the dataset
# default is in integer format,have to convert it into factor.
str(credit)

'data.frame':	1000 obs. of  21 variables:
 $ checking_balance    : Factor w/ 4 levels "< 0 DM","> 200 DM",..: 1 3 4 1 1 4 4 3 4 3 ...
 $ months_loan_duration: int  6 48 12 42 24 36 24 36 12 30 ...
 $ credit_history      : Factor w/ 5 levels "critical","delayed",..: 1 5 1 5 2 5 5 5 5 1 ...
 $ purpose             : Factor w/ 10 levels "business","car (new)",..: 8 8 5 6 2 5 6 3 8 2 ...
 $ amount              : int  1169 5951 2096 7882 4870 9055 2835 6948 3059 5234 ...
 $ savings_balance     : Factor w/ 5 levels "< 100 DM","> 1000 DM",..: 5 1 1 1 1 5 4 1 2 1 ...
 $ employment_length   : Factor w/ 5 levels "> 7 yrs","0 - 1 yrs",..: 1 3 4 4 3 3 1 3 4 5 ...
 $ installment_rate    : int  4 2 2 2 3 2 3 2 2 4 ...
 $ personal_status     : Factor w/ 4 levels "divorced male",..: 4 2 4 4 4 4 4 4 1 3 ...
 $ other_debtors       : Factor w/ 3 levels "co-applicant",..: 3 3 3 2 3 3 3 3 3 3 ...
 $ residence_history   : int  4 2 3 4 4 4 4 2 4 2 ...
 $ property            : Factor w/ 4 levels "building soci

In [12]:
# summary of the datset

summary(credit)

   checking_balance months_loan_duration                credit_history
 < 0 DM    :274     Min.   : 4.0         critical              :293   
 > 200 DM  : 63     1st Qu.:12.0         delayed               : 88   
 1 - 200 DM:269     Median :18.0         fully repaid          : 40   
 unknown   :394     Mean   :20.9         fully repaid this bank: 49   
                    3rd Qu.:24.0         repaid                :530   
                    Max.   :72.0                                      
                                                                      
       purpose        amount           savings_balance  employment_length
 radio/tv  :280   Min.   :  250   < 100 DM     :603    > 7 yrs   :253    
 car (new) :234   1st Qu.: 1366   > 1000 DM    : 48    0 - 1 yrs :172    
 furniture :181   Median : 2320   101 - 500 DM :103    1 - 4 yrs :339    
 car (used):103   Mean   : 3271   501 - 1000 DM: 63    4 - 7 yrs :174    
 business  : 97   3rd Qu.: 3972   unknown      :183    unemplo

In [13]:
#checking balance frequency table
table(credit$checking_balance)


    < 0 DM   > 200 DM 1 - 200 DM    unknown 
       274         63        269        394 

In [14]:
# savings balance frequency table
table(credit$savings_balance)


     < 100 DM     > 1000 DM  101 - 500 DM 501 - 1000 DM       unknown 
          603            48           103            63           183 

In [15]:
# Summary of loan duration
summary(credit$months_loan_duration)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
    4.0    12.0    18.0    20.9    24.0    72.0 

In [16]:
# summary of amount
summary(credit$amount)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
    250    1366    2320    3271    3972   18424 

In [17]:
# default(tager) variable frequency table
table(credit$default)


  1   2 
700 300 

From the above checking balance,saving balance are categorical variables.There are 24 records with less than 0 DM it means there is negative balance.greter than 200 DM are 63.1-200 DM are 269.394 records are unknown.
Months loan durtion an amount are numerical variables. Loan duration is between 4 to 72 months with median value of 18. Amount is between 250 to 18424 with median value of 2320DM.


## Splitting the data

I splitted the data 90% training and 10% for testing. so that the target variable is balanced in both.

In [14]:
set.seed(123)

In [15]:
# out of 1000, 900 are for training.
train_sample <- sample(1000, 900)

In [16]:
str(train_sample)

 int [1:900] 415 463 179 526 195 938 818 118 299 229 ...


In [22]:
# Remaining of training is given for testing
credit_train <- credit[train_sample, ]
credit_test <- credit[-train_sample, ]

In [18]:
prop.table(table(credit_train$default))


        1         2 
0.7055556 0.2944444 

In [19]:
prop.table(table(credit_test$default))


   1    2 
0.65 0.35 

## Model

C50 package is used for Decision tree algorithm. which used Entropy and information gain for splitting the tree.



In [10]:
install.packages("C50")

package 'C50' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\kgajj\AppData\Local\Temp\Rtmp4Ckpvl\downloaded_packages


In [23]:
# Training the model
library(C50)
credit_model <- C5.0(credit_train[,-c(17)], factor(credit_train$default))
summary(credit_model)


Call:
C5.0.default(x = credit_train[, -c(17)], y = factor(credit_train$default))


C5.0 [Release 2.07 GPL Edition]  	Wed Jun 12 12:33:39 2019
-------------------------------

Class specified by attribute `outcome'

Read 900 cases (21 attributes) from undefined.data

Decision tree:

checking_balance in {> 200 DM,unknown}: 1 (412/54)
checking_balance in {< 0 DM,1 - 200 DM}:
:...credit_history in {fully repaid,fully repaid this bank}:
    :...housing = rent: 2 (16/1)
    :   housing = for free:
    :   :...other_debtors = co-applicant: 1 (2)
    :   :   other_debtors in {guarantor,none}: 2 (12/1)
    :   housing = own:
    :   :...purpose in {business,car (used),retraining}: 1 (10/2)
    :       purpose in {domestic appliances,education,others,radio/tv,
    :       :           repairs}: 2 (6/1)
    :       purpose = car (new):
    :       :...months_loan_duration <= 22: 2 (6)
    :       :   months_loan_duration > 22: 1 (2)
    :       purpose = furniture:
    :       :...installment_pla

From the above, The size of decision tree is 42.It used 16 attributes out of 21 attributes. for tree building.Checking balance is the root node, it means the information gain from this variable is maximal compared to all.

In [24]:
# predicting for test dataset.
credit_pred <- predict(credit_model, credit_test)

In [25]:
install.packages('gmodels')
library(gmodels)
CrossTable(factor(credit_test$default), credit_pred,
prop.chisq = FALSE, prop.c = FALSE, prop.r = FALSE,
dnn = c('actual default', 'predicted default'))

package 'gmodels' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\kgajj\AppData\Local\Temp\Rtmp4Ckpvl\downloaded_packages

 
   Cell Contents
|-------------------------|
|                       N |
|         N / Table Total |
|-------------------------|

 
Total Observations in Table:  100 

 
               | predicted default 
actual default |         1 |         2 | Row Total | 
---------------|-----------|-----------|-----------|
             1 |        55 |        10 |        65 | 
               |     0.550 |     0.100 |           | 
---------------|-----------|-----------|-----------|
             2 |        22 |        13 |        35 | 
               |     0.220 |     0.130 |           | 
---------------|-----------|-----------|-----------|
  Column Total |        77 |        23 |       100 | 
---------------|-----------|-----------|-----------|

 


From above cross table,out of 65 non default 55 predicted as non default and 10 predicted wrong. Out of 35 default 13 predicted as  default. The model is not very accurate as it gave just 68% accuracy.

In [27]:
install.packages('lubridate')
library(lubridate)
library(caret)
confusionMatrix(factor(credit_test$default),factor(credit_pred))

Warning message:
"package 'lubridate' is in use and will not be installed"

Confusion Matrix and Statistics

          Reference
Prediction  1  2
         1 55 10
         2 22 13
                                          
               Accuracy : 0.68            
                 95% CI : (0.5792, 0.7698)
    No Information Rate : 0.77            
    P-Value [Acc > NIR] : 0.98560         
                                          
                  Kappa : 0.2363          
                                          
 Mcnemar's Test P-Value : 0.05183         
                                          
            Sensitivity : 0.7143          
            Specificity : 0.5652          
         Pos Pred Value : 0.8462          
         Neg Pred Value : 0.3714          
             Prevalence : 0.7700          
         Detection Rate : 0.5500          
   Detection Prevalence : 0.6500          
      Balanced Accuracy : 0.6398          
                                          
       'Positive' Class : 1               
                                    

The accuarcy of the model is 68%. the specificity is just 56%. we canimprove the model performence by boosting.

## Bossting

As the above model performence is less, we can improve performance by boosting.boosting is rooted in the notion that by combining a number of weak performing learners, we can create a team that is much stronger than any of the learners alone.

For boosting we need to add an additional trials parameter indicating the number of separate decision trees to use in the boosted team. The trials parameter sets an upper limit; the algorithm will stop adding trees if it recognizes that additional trials do not seem to be improving the accuracy.

In [88]:
# by adding trails
credit_boost10 <- C5.0(credit_train[-17], factor(credit_train$default),
trials = 10)

In [87]:
credit_boost10



Call:
C5.0.default(x = credit_train[-17], y = factor(credit_train$default), trials
 = 10)


C5.0 [Release 2.07 GPL Edition]  	Wed Jun 12 12:38:00 2019
-------------------------------

Class specified by attribute `outcome'

Read 900 cases (21 attributes) from undefined.data

-----  Trial 0:  -----

Decision tree:

checking_balance in {> 200 DM,unknown}: 1 (412/54)
checking_balance in {< 0 DM,1 - 200 DM}:
:...credit_history in {fully repaid,fully repaid this bank}:
    :...housing = rent: 2 (16/1)
    :   housing = for free:
    :   :...other_debtors = co-applicant: 1 (2)
    :   :   other_debtors in {guarantor,none}: 2 (12/1)
    :   housing = own:
    :   :...purpose in {business,car (used),retraining}: 1 (10/2)
    :       purpose in {domestic appliances,education,others,radio/tv,
    :       :           repairs}: 2 (6/1)
    :       purpose = car (new):
    :       :...months_loan_duration <= 22: 2 (6)
    :       :   months_loan_duration > 22: 1 (2)
    :       purpose = furniture

In [58]:
summary(credit_boost10)


Call:
C5.0.default(x = credit_train[-17], y = factor(credit_train$default), trials
 = 10)


C5.0 [Release 2.07 GPL Edition]  	Mon Jun 10 23:38:43 2019
-------------------------------

Class specified by attribute `outcome'

Read 900 cases (21 attributes) from undefined.data

-----  Trial 0:  -----

Decision tree:

checking_balance in {> 200 DM,unknown}: 1 (412/54)
checking_balance in {< 0 DM,1 - 200 DM}:
:...credit_history in {fully repaid,fully repaid this bank}:
    :...housing = rent: 2 (16/1)
    :   housing = for free:
    :   :...other_debtors = co-applicant: 1 (2)
    :   :   other_debtors in {guarantor,none}: 2 (12/1)
    :   housing = own:
    :   :...purpose in {business,car (used),retraining}: 1 (10/2)
    :       purpose in {domestic appliances,education,others,radio/tv,
    :       :           repairs}: 2 (6/1)
    :       purpose = car (new):
    :       :...months_loan_duration <= 22: 2 (6)
    :       :   months_loan_duration > 22: 1 (2)
    :       purpose = furniture

In [30]:
credit_boost_pred10 <- predict(credit_boost10, credit_test)


In [31]:
CrossTable(credit_test$default, credit_boost_pred10,
prop.chisq = FALSE, prop.c = FALSE, prop.r = FALSE,
dnn = c('actual default', 'predicted default'))


 
   Cell Contents
|-------------------------|
|                       N |
|         N / Table Total |
|-------------------------|

 
Total Observations in Table:  100 

 
               | predicted default 
actual default |         1 |         2 | Row Total | 
---------------|-----------|-----------|-----------|
             1 |        59 |         6 |        65 | 
               |     0.590 |     0.060 |           | 
---------------|-----------|-----------|-----------|
             2 |        17 |        18 |        35 | 
               |     0.170 |     0.180 |           | 
---------------|-----------|-----------|-----------|
  Column Total |        76 |        24 |       100 | 
---------------|-----------|-----------|-----------|

 


with boosting total numbee of non default preicted increased to 59 and default predicted increased to 18.eventually the performence got increased.

In [32]:
confusionMatrix(factor(credit_test$default),factor(credit_boost_pred10))

Confusion Matrix and Statistics

          Reference
Prediction  1  2
         1 59  6
         2 17 18
                                          
               Accuracy : 0.77            
                 95% CI : (0.6751, 0.8483)
    No Information Rate : 0.76            
    P-Value [Acc > NIR] : 0.46145         
                                          
                  Kappa : 0.455           
                                          
 Mcnemar's Test P-Value : 0.03706         
                                          
            Sensitivity : 0.7763          
            Specificity : 0.7500          
         Pos Pred Value : 0.9077          
         Neg Pred Value : 0.5143          
             Prevalence : 0.7600          
         Detection Rate : 0.5900          
   Detection Prevalence : 0.6500          
      Balanced Accuracy : 0.7632          
                                          
       'Positive' Class : 1               
                                    

The  accuracy increased to 77%. The specificty is increased to 75%.

## Checking model with penalizing.

The business has given false negative has a cost of 4 versus a false positive's cost of 1. SO penalized those values with error cost.

In [47]:
matrix_dimensions <- list(c("1", "2"), c("1", "2"))
names(matrix_dimensions) <- c("predicted", "actual")

In [48]:
matrix_dimensions

$predicted
[1] "1" "2"

$actual
[1] "1" "2"

In [81]:
error_cost <- matrix(c(0, 1, 4, 0), nrow = 2,
dimnames = matrix_dimensions)

In [82]:
error_cost

actual
predicted 1 2
        1 0 4
        2 1 0

In [83]:
credit_cost <- C5.0(credit_train[,-c(17)], factor(credit_train$default),
costs = error_cost,na.action = na.pass)


In [84]:
summary(credit_cost)


Call:
C5.0.default(x = credit_train[, -c(17)], y =
 factor(credit_train$default), costs = error_cost, na.action = na.pass)


C5.0 [Release 2.07 GPL Edition]  	Wed Jun 12 13:09:49 2019
-------------------------------

Class specified by attribute `outcome'

Read 900 cases (21 attributes) from undefined.data
Read misclassification costs from undefined.costs

Decision tree:

checking_balance = unknown:
:...installment_plan in {bank,stores}:
:   :...employment_length in {0 - 1 yrs,1 - 4 yrs,unemployed}: 2 (30/16)
:   :   employment_length = 4 - 7 yrs: 1 (10)
:   :   employment_length = > 7 yrs:
:   :   :...age <= 41: 2 (9/5)
:   :       age > 41: 1 (12)
:   installment_plan = none:
:   :...credit_history in {fully repaid,fully repaid this bank}: 1 (4)
:       credit_history = critical:
:       :...amount <= 11760: 1 (118/1)
:       :   amount > 11760: 2 (1)
:       credit_history = delayed:
:       :...installment_rate > 3: 2 (15/9)
:       :   installment_rate <= 3:
:       :   :...resid

In [85]:
credit_cost_pred <- predict(credit_cost, credit_test)


In [86]:
CrossTable(credit_test$default, credit_cost_pred,
prop.chisq = FALSE, prop.c = FALSE, prop.r = FALSE,
dnn = c('actual default', 'predicted default'))


 
   Cell Contents
|-------------------------|
|                       N |
|         N / Table Total |
|-------------------------|

 
Total Observations in Table:  100 

 
               | predicted default 
actual default |         1 |         2 | Row Total | 
---------------|-----------|-----------|-----------|
             1 |        43 |        22 |        65 | 
               |     0.430 |     0.220 |           | 
---------------|-----------|-----------|-----------|
             2 |         8 |        27 |        35 | 
               |     0.080 |     0.270 |           | 
---------------|-----------|-----------|-----------|
  Column Total |        51 |        49 |       100 | 
---------------|-----------|-----------|-----------|

 


## Decision Rule Making on Mashroom Dataset

## Data Extraction

Data extracted from Machine learning with R Github.It has 8124 records with 23 variables. The target variable is type whether it is e(edible) or p(poisonous) of mashroom with given features like shape, surface, color etc. all of them are factor variables.

In [92]:
mushrooms <- read.csv("C:\\Users\\kgajj\\AppData\\Local\\Temp\\Temp1_Machine-Learning-with-R-datasets-master.zip\\Machine-Learning-with-R-datasets-master\\mushrooms.csv", stringsAsFactors = TRUE)

In [93]:
str(mushrooms)

'data.frame':	8124 obs. of  23 variables:
 $ type                    : Factor w/ 2 levels "e","p": 2 1 1 2 1 1 1 1 2 1 ...
 $ cap_shape               : Factor w/ 6 levels "b","c","f","k",..: 6 6 1 6 6 6 1 1 6 1 ...
 $ cap_surface             : Factor w/ 4 levels "f","g","s","y": 3 3 3 4 3 4 3 4 4 3 ...
 $ cap_color               : Factor w/ 10 levels "b","c","e","g",..: 5 10 9 9 4 10 9 9 9 10 ...
 $ bruises                 : Factor w/ 2 levels "f","t": 2 2 2 2 1 2 2 2 2 2 ...
 $ odor                    : Factor w/ 9 levels "a","c","f","l",..: 7 1 4 7 6 1 1 4 7 1 ...
 $ gill_attachment         : Factor w/ 2 levels "a","f": 2 2 2 2 2 2 2 2 2 2 ...
 $ gill_spacing            : Factor w/ 2 levels "c","w": 1 1 1 1 2 1 1 1 1 1 ...
 $ gill_size               : Factor w/ 2 levels "b","n": 2 1 1 2 1 1 1 1 2 1 ...
 $ gill_color              : Factor w/ 12 levels "b","e","g","h",..: 5 5 6 6 5 6 3 6 8 3 ...
 $ stalk_shape             : Factor w/ 2 levels "e","t": 1 1 1 1 2 1 1 1 1 1 ...
 $ stalk_r

 The variable Veil type has same level for all the types. so it does not make any affect on the model. so i deleted it.

In [94]:
# deleting one column
mushrooms$veil_type <- NULL

In [95]:
# checking the class variables.how many are  e type or p type.
table(mushrooms$type)


   e    p 
4208 3916 

More than 50% of the mashrooms are edible and remaining are poisonous.

Since ZeroR ignores all of the features and simply predicts the target's mode.its rule would state that all the mushrooms are edible. Obviously,this is not a very helpful classifier, because it would leave a mushroom gatherer sick or dead. We will use the 1R implementation in the RWeka package called OneR().



In [96]:
install.packages("RWeka")
library(RWeka)
mushroom_1R <- OneR(type ~ ., data = mushrooms)

Warning message:
"package 'RWeka' is in use and will not be installed"

In [97]:
mushroom_1R

odor:
	a	-> e
	c	-> p
	f	-> p
	l	-> e
	m	-> p
	n	-> e
	p	-> p
	s	-> p
	y	-> p
(8004/8124 instances correct)


In [98]:
summary(mushroom_1R)


=== Summary ===

Correctly Classified Instances        8004               98.5229 %
Incorrectly Classified Instances       120                1.4771 %
Kappa statistic                          0.9704
Mean absolute error                      0.0148
Root mean squared error                  0.1215
Relative absolute error                  2.958  %
Root relative squared error             24.323  %
Total Number of Instances             8124     

=== Confusion Matrix ===

    a    b   <-- classified as
 4208    0 |    a = e
  120 3796 |    b = p

Considering that the learner utilized only a single feature, it did reasonably well.if one avoids unappetizing smells when foraging for mushrooms, they will almost avoid a trip to the hospital

## Improving Model Performence

the process of training a JRip() model is very similar to how we previously trained a OneR() model. This is one of the pleasant benefits of the functions in the RWeka package; the syntax is consistent across algorithms, which makes the process of comparing a number of different models very simple.

In [99]:
mushroom_JRip <- JRip(type ~ ., data = mushrooms)

In [100]:
mushroom_JRip

JRIP rules:

(odor = f) => type=p (2160.0/0.0)
(gill_size = n) and (gill_color = b) => type=p (1152.0/0.0)
(gill_size = n) and (odor = p) => type=p (256.0/0.0)
(odor = c) => type=p (192.0/0.0)
(spore_print_color = r) => type=p (72.0/0.0)
(stalk_surface_below_ring = y) and (stalk_surface_above_ring = k) => type=p (68.0/0.0)
(habitat = l) and (cap_color = w) => type=p (8.0/0.0)
(stalk_color_above_ring = y) => type=p (8.0/0.0)
 => type=e (4208.0/0.0)

Number of Rules : 9


In [101]:
summary(mushroom_JRip)


=== Summary ===

Correctly Classified Instances        8124              100      %
Incorrectly Classified Instances         0                0      %
Kappa statistic                          1     
Mean absolute error                      0     
Root mean squared error                  0     
Relative absolute error                  0      %
Root relative squared error              0      %
Total Number of Instances             8124     

=== Confusion Matrix ===

    a    b   <-- classified as
 4208    0 |    a = e
    0 3916 |    b = p

The JRip() classifier learned a total of nine rules from the mushroom data. An easy way to read these rules is to think of them as a list of if-else statements, similar to programming logic. The first three rules could be expressed as:
• If the odor is foul, then the mushroom type is poisonous
• If the gill size is narrow and the gill color is buff, then the mushroom type is poisonous
• If the gill size is narrow and the odor is pungent, then the mushroom type is poisonous.
• Else, the mushroom is edible

## Summary

The two classification methods that use so-called "greedy" algorithms to partition the data according to feature values. Decision trees use a divide and conquer strategy to create flowchart-like structures, while rule learners separate and conquer data to identify logical if-else rules. Both methods produce models that can be interpreted without a statistical background.

### Credit dataset:

The C5.0 algorithm to create a tree to predict whether a loan applicant will default. Using options for boosting and cost-sensitive errors, we were able to improve our accuracy and avoid risky loans that would cost the bank more money.

### Mashroom Datset:

The two rule learners, 1R and RIPPER, to develop rules to identify poisonous mushrooms. The 1R algorithm used a single feature to achieve 99 percent accuracy in identifying potentially fatal mushroom samples. On the other hand, the set of nine rules generated by the more sophisticated RIPPER algorithm correctly identified the edibility of each mushroom.
